flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/AbstractHaJobRunITCase.java

```diff
@@ -0,0 +1,133 @@
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import javax.annotation.Nullable;
+
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * {@code AbstractHaJobRunITCase} runs a job storing in HA mode and provides {@code abstract}
+ * methods for initializing a specific {@link FileSystem}.
+ */
+@ExtendWith(TestLoggerExtension.class)
+public abstract class AbstractHaJobRunITCase {
+
+    @RegisterExtension
+    private static final AllCallbackWrapper<ZooKeeperExtension> ZOOKEEPER_EXTENSION =
+            new AllCallbackWrapper<>(new ZooKeeperExtension());
+
+    @RegisterExtension
+    public final EachCallbackWrapper<MiniClusterExtension> miniClusterExtension =
+            new EachCallbackWrapper<>(
+                    new MiniClusterExtension(
+                            new MiniClusterResourceConfiguration.Builder()
+                                    .setNumberTaskManagers(1)
+                                    .setNumberSlotsPerTaskManager(1)
+                                    .setRpcServiceSharing(RpcServiceSharing.DEDICATED)
+                                    .setConfiguration(getFlinkConfiguration())
+                                    .build()));
+
+    private Configuration getFlinkConfiguration() {
+        Configuration config = new Configuration();
+        config.set(HighAvailabilityOptions.HA_MODE, "ZOOKEEPER");
+        config.set(
+                HighAvailabilityOptions.HA_ZOOKEEPER_QUORUM,
+                ZOOKEEPER_EXTENSION.getCustomExtension().getConnectString());
+
+        config.set(HighAvailabilityOptions.HA_STORAGE_PATH, createHAStoragePath());
+
+        updateConfiguration(config);
+
+        FileSystem.initialize(config, loadPluginManager());
+        initializeStorageBackend(config);
+
+        return config;
+    }
+
+    /**
+     * Should return the path to the HA storage which will be injected into the Flink configuration.
+     *
+     * @see HighAvailabilityOptions#HA_STORAGE_PATH
+     */
+    protected abstract String createHAStoragePath();
+
+    /**
+     * Updates the passed {@link Configuration} to point to the {@link FileSystem} that's subject to
+     * test.
+     */
+    protected abstract void updateConfiguration(Configuration config);
+
+    /** Runs any additional initialization that are necessary before running the actual test. */
+    protected void initializeStorageBackend(Configuration config) {}
+
+    @Nullable
+    protected PluginManager loadPluginManager() {
+        // by default, no PluginManager is loaded
+        return null;
+    }
+
+    @Test
+    public void testJobExecutionInHaMode() throws Exception {
+        final MiniCluster flinkCluster = miniClusterExtension.getCustomExtension().getMiniCluster();
+
+        final JobGraph jobGraph = JobGraphTestUtils.singleNoOpJobGraph();
+
+        flinkCluster.runDetached(jobGraph);
```

Comment:
In case of the `FileSystemJobResultStore` issue covered by this PR, this test would run forever because the error appeared during the initialization. I didn't find a better way to assert the state of the `MiniCluster` here. Checking the `isRunning()` method wouldn't help because it's switched immediately into state running.

IMHO, we should make the FatalErrorHandler available through the `MiniClusterResource`. Right now, it's instantiated internally calling a `closeAsync()` on the `MiniCluster`. Making this one accessible in the test would help adding an assert on that. But it felt like a separate change.

Reply:
I'm planning to create a ticket to cover this functionality after my current approach was reviewed to make sure that I don't miss anything

Reply:
@zentol I would like to get your opinion on this issue. Would you agree that exposing the `FatalErrorHandler` for testing purposes is a valid improvement?

Reply:
That'd be fine I think.

Reply:
The issue is that MiniCluster is running in a separate thread. It fails while I'm submitting the Job which keeps the job submission to never complete. But using the non-detached job submission is a good point: I added a timeout to that to make the test fail earlier. That way, the test would run forever if there was an initialization error.

Initially, my idea was that I could wait for the `FatalErrorHandler` (if it would be exposed outside of the MiniCluster) to report an error and fail in that case. Waiting for a reasonable long time with no error occurring would have been an indication that the initialization of the MiniCluster succeeded and I could have continued with the test.

But that approach has the downside that the test has to wait in cases where it actually succeeds which is bad in terms of test runtime. Making the job submission fail after a specific timeout sounds like the better approach.